algebra-core/src/fields/mod.rs

```diff
@@ -193,6 +195,14 @@ pub trait FpParameters: 'static + Send + Sync + Sized {
 
     /// (Self::MODULUS - 1) / 2
     const MODULUS_MINUS_ONE_DIV_TWO: Self::BigInt;
+
+    /// The base of a small subgroup that can be used for mixed-radix FFT.
+    const SMALL_SUBGROUP_BASE: Option<u32> = None;
+    /// The power of a small subgroup that can be used for mixed-radix FFT.
+    const SMALL_SUBGROUP_POWER: Option<u32> = None;
+    /// GENERATOR^((MODULUS-1) / (2^s *
+    /// SMALL_SUBGROUP_BASE^SMALL_SUBGROUP_POWER)) Used for mixed-radix FFT.
+    const FULL_ROOT_OF_UNITY: Option<Self::BigInt> = None;
```

Comment:
As I mentioned on Telegram, maybe we should move these parameters, along with the `ROOT_OF_UNITY`-related stuff to a separate `FftField: Field` trait?

Then we require `PrimeField: FftField`?

This way we can change the EvaluationDomain to only care about the FFT structure, so that it becomes `EvaluationDomain<F: FftField>`.

Reply:
Would you put `TWO_ADICITY` also into the `FftField` trait?
As far as I can see, it is mostly used in the FFT related stuff with the exception of one occurrence in the `sqrt_impl`. But then, the `sqrt_impl` also implicitly relies on the `ROOT_OF_UNITY` (which is the default for `qnr_to_t`).

Reply:
Also, I think there are three options here:
1. Putting the methods and constants into the `FftField` trait, which requires to implement this trait for our prime fields (just as we do with FpParameters right now).
2. Putting only the methods into the `FftField` trait, so that the `FftField` trait can be implemented in the `impl_Fp` macro.
3. Actually having a second trait like `FftParameters`, which has the constants and needs to be implemented alongside `FpParameters`. The `FftField` trait then works analogously to the `PrimeField` trait in that it has a `Params` type. The methods of the `FftField` trait can then be implemented in the `impl_Fp` macro.

The disadvantage of the first option is that methods like `full_root_of_unity` cannot have a default implementation as there is no way of creating a `Self` from just a `Self::BigInt` at this point.

Reply:
I think option 3 is the way to go, though it is a little bit more work than the other options. We can make `FpParameters: FftParameters`, and `PrimeField: FftField`, with the appropriate bound on the Parameters associated type

Reply:
And yeah, I’d put TWO_ADICITY and related constants in the new parameters trait

Reply:
It turns out that splitting these traits is not easily achievable without any downsides.

The problem arises from the following constraints:
1. For the ff-fft crate to be able to work, it requires `F: FftField` and `FftField` must have a type parameter `FftParams: FftParameters`.
2. We would like that `PrimeField: FftField`. Since, `PrimeField` requires a type parameter `Params: FpParameters`, we end up with something similar to the following:
```rust
trait FftParameters {
    type BigInt: BigInteger;
    ...
}
trait FpParameters: FftParameters { ... }
trait FftField: Field {
    type FftParams: FftParameters;
}
trait PrimeField: FftField {
     type Params: FpParameters<BigInt = Self::BigInt>;
     type BigInt: BigInteger;
}
```
This still works fine. However, we would like to require the two BigInts to be the same or even `FftParams = Params` for the `PrimeField`.
Doing something like `trait PrimeField: FftField<FftParams = Self::Params>`, results in a cyclic dependency though ([error[E0391]](https://doc.rust-lang.org/error-index.html#E0391)).

So far, I haven't found a satisfactory solution splitting these traits. So, I'm not sure this is worth it unless you have another idea that I might be missing here. :)

Reply:
I think if you do `PrimeField: FftField<FftParams = <Self as PrimeField::Params>>`, then it should work?

Reply:
You're completely right. Perfect, then this will work out. :)